### ISSUE TITLE
GoMetalinter not reporting test or testify linter failures

### ISSUE THREAD

**Issue by tommyknows** [fatih/vim-go#2131]
### What did you do?
I have the following lines in my `.vimrc`:
```
let g:go_metalinter_autosave = 1
let g:go_metalinter_enabled = ['vet', 'golint', 'errcheck', 'test', 'testify']
let g:go_metalinter_deadline = "5s"
let g:go_metalinter_autosave_enabled = ['vet', 'golint', 'errcheck', 'test', 'testify']
```
Executing `:let go_metalinter_autosave_enabled` gives me `go_metalinter_autosave_enabled ['vet', 'golint', 'errcheck', 'test', 'testify']`, so this seems to work.

### What did you expect to happen?
I have an error in my tests, which would mean that Metalinter should fail when I call it (or save the file).

### What happened instead?
When saving the file or executing `:GoMetaLinter`, it returns `SUCCESS`.
When executing `:GoMetaLinter -E 'test' -E 'testify'`, it fails (as it should)

### Note
Calling Metalinter from the command line also works just fine

### Configuration (**MUST** fill this out):

* vim-go version:
master (using plug and ran `:PlugUpdate`)

* `vimrc` you used to reproduce (use a *minimal* vimrc with other plugins disabled; do not link to a 2,000 line vimrc):
(only pasting the go-related stuff):
My vimrc is quite long (~500 lines). If there's an indication that it is the vim-config, i'll be happy to shorten it.

* Vim version (first three lines from `:version`):
VIM - Vi IMproved 8.1 (2018 May 18, compiled Dec 28 2018 20:12:37)
macOS version
Included patches: 1-650

* Go version (`go version`):
go version go1.11.1 darwin/amd64

**Comment by bhcleek** [comment]
From the docs:
> Use this option to auto :GoMetaLinter on save. Only linter messages for
the active buffer will be shown.

**Comment by tommyknows** [comment]
Even when having only the `*_test.go` file opened (or the normal file), it does not work. still get a `SUCCESS` message.

**Comment by bhcleek** [comment]
I'm having a difficult time fully duplicating the behavior you're describing. All my attempts to get `gometalinter` to fail using the `test` linter have failed. What do you get from the command line for `gometalinter --disable-all --linter=test -t` for the package you're expecting to fail?

**Comment by bhcleek** [comment]
With these settings, I see the `errcheck` results after writing the buffer:
```
let g:go_metalinter_autosave = 1
let g:go_metalinter_enabled = ['vet', 'golint', 'errcheck', 'test', 'testify']
let g:go_metalinter_deadline = "5s"
let g:go_metalinter_autosave_enabled = ['vet', 'golint', 'errcheck', 'test', 'testify']
```

Do you?

**Comment by tommyknows** [comment]
Executing `gometalinter --disable-all --linter=test -t` from the command line gives me
```
gometalinter: error: expected KEY=VALUE got 'test', try --help
```

If you meant something like this:
```
> gometalinter --disable-all --enable=test --enable=testify -t
board_test.go:198::error: Not equal (testify)
```

However, with the `let` statements (they're also already in my vimrc), I cannot get gometalinter to fail.
Snippet from the go code:
```go
if colIdx != board.colIdx {
    t.Errorf("Not Equal!")
}
assert.Equal(t, colIdx, board.colIdx)
```

What's interesting is that only `testify` gives an error, while the `test` linter does not. 
Running `go test` however gives me both errors (one from the `t.Errorf` and one from `assert.Equal`).

Basically I cannot get gometalinter to fail when called over the wrapper, it works just fine from the CLI... I'd be happy to dig into it, but I don't know where I should start to look.

Also, I'm just talking about the `test` and `testify` linters. If I have any syntax-errors, the other linters work and come up in the quickfix list (golint, errcheck, vet).

**Comment by bhcleek** [comment]
> If you meant something like this:

I did.

> I'd be happy to dig into it, but I don't know where I should start to look.

First thing to try is to get the exact command line that's vim-go is using to run `gometalinter` and run it from the command line to see what `gometalinter` is outputting. You can get the command by using `let g:go_debug=['shell-commands']` and then running `:messages` to see what it ran.

vim-go's linting code is in `autoload/go/lint.vim`.

**Comment by bhcleek** [comment]
I'm able to duplicate what you're seeing with the testify error not being shown. I"ll work to fix that.
 
It seems, though, that `gometalinter`'s `test` linter doesn't at all  work as expected.

**Comment by bhcleek** [comment]
I can't get `gometalinter`'s `test` linter to work as expected at the commandline.

I was able to show that its `testify` linter works as expected at the commandline and that vim-go reports its results as expected, though:

![2131](https://user-images.githubusercontent.com/1958864/51796721-478c9d80-21ac-11e9-98ab-cfc5a64975db.gif)

**Comment by tommyknows** [comment]
I can confirm that the command that is executed is correct (linters are active).
But I can't get gometalinter to fail with just `testify` enabled. Have you made any changes to the code?

Thanks for the help!

**Comment by bhcleek** [comment]
I haven't made any changes to the code.

**Comment by bhcleek** [comment]
> But I can't get gometalinter to fail with just testify enabled

Do you mean at the commandline with `gometalinte` or in vim-go with `:GoMetaLinter`?

**Comment by tommyknows** [comment]
Just with `vim-go` / in vim itself. Commandline, as described [here](https://github.com/fatih/vim-go/issues/2131#issuecomment-457872758) is working fine.

**Comment by palsivertsen** [comment]
FYI: This might be a problem with `gometalinter`. It does not play well with the multiline outputs given by the `go test` tool and `testify` package. I did some improvements in [this](https://github.com/alecthomas/gometalinter/pull/489) pull request, but it's far from perfect. For example a panic will be silently ignored.

IIRC `gometalinter` does regex marches per line which makes it difficult to handle errors where key information is spread over multiple lines.

To confirm that this is a `gometaliner` issue you could:
1. Verify that the tests fail by running `go test`
2. Check if `gometalinter` picks up the error(s):
  Run `gometalinter --disable-all --enable=test`
  **AND/OR**
  Run `gometalinter --disable-all --enable=testify`

If testify can not pick up the error, then there's your issue

**Comment by tommyknows** [comment]
Sorry for the late reply.
Yes, tests are failing, and yes, gometalinter picks up the issues. It's the `testify` linter that fails in my case (as it is an `assert.Equal` operation for example). 
Output of `gometalinter --disable-all --enable=testify`:
```
gometalinter --disable-all --enable=testify
board_test.go:223::error: Not equal (testify)
board_test.go:223::error: Not equal (testify)
board_test.go:223::error: Not equal (testify)
[...]
```

Note: multiple lines output as I use table-driven tests and every entry fails.

**Comment by bhcleek** [comment]
That output looks correct and appears to conform to vim-go expectations. `:GoMetaLinter` continues to pick up testify errors correctly for me. Can you verify the output of `:echo g:go_metalinter_enabled`?

**Comment by tommyknows** [comment]
The command returns
```
['vet', 'golint', 'errcheck', 'test', 'testify']
```

I also tested with only having [`testify`] or  [`testify`, `test`] enabled.
Also set the deadline of metalinter to `50s`, but metalinter prints its success quite fast (~1s).

**Comment by qjvtenkroode** [comment]
I'm also having this issue, not using testify but only the test linter. Changing my vimrc to just use: `let g:go_metalinter_autosave_enabled = ['test'] ` does nothing. Checking the command run by the vim gives:
`gometalinter --disable-all --tests --enable=test --include=^myfile.go:.*$ <locationtomyfile.go>`

Now running this with `-d` to enable gometalinter debug shows that the failing tests are printed as a  WARN
```
DEBUG: [Mar 13 19:30:38.728] [test.1]: warning: /usr/local/bin/go returned exit status 1: --- FAIL: TestParse (0.00s)
    conto_test.go:40: assertion failed: 
        --- ret1
        +++ expected1
        {conto.Entry}.ID:
                -: "48328b4020a7e3e6045a775963d912b6"
                +: "q48328b4020a7e3e6045a775963d912b6"
        
FAIL
FAIL    github.com/qjvtenkroode/conto   0.007s

DEBUG: [Mar 13 19:30:38.728] [test.1]: test hits 0: (?m:^\t(?P<path>.*?\.go):(?P<line>\d+): (?P<message>.+)$)
DEBUG: [Mar 13 19:30:38.728] [test.1]: test linter took 676.885976ms
DEBUG: [Mar 13 19:30:38.728] total elapsed time 681.483468ms
```

This seems to suggest that @palsivertsen is correct about gometalinter failing on multiline output.

**Comment by bhcleek** [comment]
gometalinter is deprecated and is going to be archived in a couple of weeks, and we are working on support for golangci-lint to take it's place in vim-go.

Since gometalinter is being deprecated, we won't be putting any effort to trying to resolve this.
